### DISCUSSION TITLE
Timeout Exceeded Error When Grading Questions

### DISCUSSION THREAD

**Discussion by aalrabah** [PrairieLearn/PrairieLearn#10082] [Dev]
I'm running a Docker container locally and testing some questions in PrairieLearn. On a homework page with 15 questions, 10 of them aren't working. When I check the terminal logs, I see this message:

info: Timeout exceeded; killing container

<img width="898" alt="Screenshot 2024-06-26 at 11 50 18 AM" src="https://github.com/PrairieLearn/PrairieLearn/assets/94386936/f3cb147c-dcba-48c9-ada9-0f8600e2a45b">

I need to fix this issue to make all questions work.

Attached is a picture of the error. The image indicates that I should not exceed a 20-second grading time limit. Even though the answer is correct, it’s still not working. Some questions occasionally exceed the time limit but still work.

How can I resolve this issue?

**Comment by nwalters512** [comment]
Can you share more details about the code you're trying to grade? Is there perhaps an infinite loop or some very expensive computation that it's doing? Describing the code, the language, and which grader image you're using would be helpful, but it would also be great if you could share the relevant parts of the code that's being graded and/or doing the grading.

**Comment by jonatanschroeder** [comment]
I'm sorry if this sounds harsh but:

http://www.catb.org/~esr/faqs/smart-questions.html#beprecise
https://www.chiark.greenend.org.uk/~sgtatham/bugs.html

**Comment by aalrabah** [comment]
The directory of the code: serverFilesCourse/grader/base_grader.py
code that is doing the grading: 

`class BaseGrader:
    # ... [other parts of the class]

    def __init__(self, grading_function, submission_num):
        self.grading_function = grading_function
        self.submission_num = submission_num
        self.submission_count = 0
        self.grading_result = dict()
        self.tests_list = []

    def run(self):
        while self.submission_count < self.submission_num:
            print("[Base Grader] Number of current test run: {}...".format(self.submission_count + 1))
            self.grading_function(self)  # Inject self instance into grading function
            self.submission_count += 1
        print("Finished Loop. Calling dump_pl_grading_result to dump json result file to PL.")
        self.dump_pl_grading_result()
`

The code I'm trying to grade is a set of SQL queries submitted by students via PrairieLearn. The language used is SQL, and the grader image in use is prairielearn/prairielearn.

Details:
Grading Environment: The grading logic is implemented in a custom Python script that interacts with a MySQL database to execute the queries and compare the results against the expected solution.
Grader Functionality: The grader function sets up the database, executes the student's query, and then compares the output with the solution query. It also checks for differences and constructs detailed feedback.

Timeout Exceeded(The error): The primary issue observed is that certain questions are being graded, leading to a timeout.(Timeout exceeded; killing container).

please let me know if you need more details about this issue.

**Reply by jonatanschroeder** [reply]
The `prairielearn/prairielearn` image is probably not a good idea for a grading image, as it is pretty large and has significantly more data than what you actually need for grading. Since your grading code is in `serverFilesCourse` you may be able to create a custom image that starts from a smaller image (e.g., [`python:3.12`](https://hub.docker.com/_/python)) and then includes any other dependencies (e.g., mysql), as well as the code you currently have in `serverFilesCourse`. This would speed up the loading of the image. We're happy to help you with this if you need additional info.

As for the reasons for the timeout itself: have you tried to debug to figure out how long each step of the running process takes, from initialization to start of unit tests and then the tests themselves? If what is causing a long time is the initialization (e.g., setting up the database), you may be able to reduce some of this time by initializing in the creation of the Docker image (Dockerfile).

If the code that you are running is indeed taking 20+ seconds in the code that actually runs the tests, and this is expected (the queries indeed take that long), then you should be able to change the timeout in the question's info.json, though you should remember that this increase means that it may take longer for students to get feedback about issues like infinite loops, which can be frustrating for students and use resources beyond what is actually necessary.

**Reply by nwalters512** [reply]
I'd echo all this: a) use a better, smaller image and b) spend some time trying to understand why things are slow before trying to increase the execution timeout, as this will result in a far better experience for students.

**Reply by aalrabah** [reply]
Thank you for your detailed feedback and suggestions. 

Given that my grading code is located in serverFilesCourse, I agree that creating a custom image starting from a smaller base image like python:3.12 and including the necessary dependencies (e.g., MySQL) would be beneficial. However, I could use some assistance in setting up this smaller Docker image correctly.

Current Setup
My current Dockerfile is one level higher than serverFilesCourse, and this is the code that I used to open up the server:


```
FROM ubuntu:20.04

# Set environment variables to non-interactive mode
ENV DEBIAN_FRONTEND=noninteractive

# Update package lists and install MySQL server, Python, and necessary packages
RUN apt-get update && apt-get install -y \
    mysql-server \
    python3-pip \
    python3-dev \
    sudo \
    && rm -rf /var/lib/apt/lists/*

# Upgrade pip and install Python packages
RUN python3 -m pip install --upgrade pip && \
    python3 -m pip install \
    sqlparse \
    mysql-connector-python \
    rstr

# Create necessary directories
RUN mkdir -p /grade/run/bin /grade/results /grade/student /grade/serverFilesCourse/sql-grader /grade/tests /grade/serverFilesCourse/grader

# Copy the run.sh script into the container
COPY serverFilesCourse/sql-grader/run.sh /usr/local/bin/run.sh

# Ensure the script has execution permissions
RUN chmod +x /usr/local/bin/run.sh

# Copy the server files into the container
COPY serverFilesCourse/ /grader/

# Expose MySQL port
EXPOSE 3306

# Start MySQL and run the grading script
CMD ["bash", "/usr/local/bin/run.sh"]

```


I then ran this to open up the server:

```
docker run -it --rm -p 3000:3000 \
    -v "$HOME/local-directory-to-the-course" \
    -v "$HOME/pl_ag_jobs:/jobs" \
    -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" \
    -v /var/run/docker.sock:/var/run/docker.sock \
    --platform linux/amd64 \
    --add-host=host.docker.internal:host-gateway \
    prairielearn/prairielearn
```

After reading your suggestion about using a smaller image, I modified the Dockerfile to:

```
FROM python:3.12-slim

# Set environment variables to non-interactive mode
ENV DEBIAN_FRONTEND=noninteractive

# Install necessary dependencies
RUN apt-get update && apt-get install -y \
    default-mysql-server \
    sudo \
    && rm -rf /var/lib/apt/lists/*

# Upgrade pip and install Python packages
RUN python3 -m pip install --upgrade pip && \
    python3 -m pip install \
    sqlparse \
    mysql-connector-python \
    rstr

# Create necessary directories
RUN mkdir -p /grade/run/bin /grade/results /grade/student /grade/serverFilesCourse/sql-grader /grade/tests /grade/serverFilesCourse/grader

# Copy the run.sh script into the container
COPY serverFilesCourse/sql-grader/run.sh /usr/local/bin/run.sh

# Ensure the script has execution permissions
RUN chmod +x /usr/local/bin/run.sh

# Copy the server files into the container
COPY serverFilesCourse/ /grader/

# Expose MySQL port
EXPOSE 3306

# Start MySQL and run the grading script
CMD ["bash", "/usr/local/bin/run.sh"]
```
And ran this in the terminal:

```
docker run -it --rm -p 3000:3000 \
    -v "$HOME/local-directory-to-the-course" \
    -v "$HOME/pl_ag_jobs:/jobs" \
    -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" \
    -v /var/run/docker.sock:/var/run/docker.sock \
    --platform linux/amd64 \
    --add-host=host.docker.internal:host-gateway \
    custom-grader
```
This didn't open up the PrairieLearn local server host as expected. I might have been doing it wrong or something. Can you please provide more help on the steps on how I can use an alternative smaller image for the Docker?

**Reply by nwalters512** [reply]
It's important to keep in mind the difference between `prairielearn/prairielearn`, which runs the PrairieLearn web server, and the custom grader image, which PrairieLearn itself will use to create a Docker container when grading. If you're running `docker run ...` to bring up a PrairieLearn web server, you should always do that with the `prairielearn/prairielearn` image. To change the image used to grade a particular question, see the `image` property within `externalGradingOptions` in your question's JSON file. See the [external grading documentation](https://prairielearn.readthedocs.io/en/latest/externalGrading/) for more details.

**Reply by jonatanschroeder** [reply]
One added comment: note that the `/grade` directory will be created (mounted) by PrairieLearn itself, so you don't need to add any files to that directory. In fact you shouldn't, since they will be overwritten. If you want any files from `serverFilesCourse` to be added to the image in `/grade/serverFilesCourse`, [you should add them to info.json](https://prairielearn.readthedocs.io/en/latest/externalGrading/#configuring-and-enabling-external-grader-support), if you haven't done so yet. You can, of course, still add them to any directory outside `/grade` (like you have done with the `run.sh` script). As for `/grade/results`, you will typically have to create it yourself in your grading script.

> ```
> # Expose MySQL port
> EXPOSE 3306
> ```

I don't think this is necessary, assuming you will be accessing the mysql database from the same host that is running the DB itself, though please test it out as I may be wrong.

**Reply by aalrabah** [reply]
Thank you very much for your support @nwalters512 @jonatanschroeder that was helpful :)
